mlir/lib/Catalyst/Transforms/QnodeToAsyncPatterns.cpp

```diff
@@ -31,6 +31,135 @@ namespace catalyst {
+        //  We just need to place awaits just before every use of the original value.
+        //  For example, take the following piece of code
+        //
+        //     %quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %some_val = some.op %quantum_results
+        //
+        //  First we translate the call to an async.execute (which was performed before this
+        //  function was called) When this function was called, the transformation is not yet
+        //  complete. Because there are still uses of %quantum_results.
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %some_val = some.op %quantum_results
+        //
+        //  This function will perform the following change:
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %await_quantum_results = async.await %async_quantum_results
+        //     %some_val = some.op %await_quantum_results
+        //
+        //  And it will do so for every use of %quantum_results
+        //  E.g.
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %await_quantum_results_0 = async.await %async_quantum_results
+        //     %some_val_0 = some.op %await_quantum_results_0
+        //     %await_quantum_results_1 = async.await %async_quantum_results_1
+        //     %some_val_1 = some.op %await_quantum_results_1
+        //
+        //  However, we can do a bit better and avoid this extra awaits that are
+        //  dominated by the first await. Keeping the awaits is not incur a large increase
+        //  in run time. Removing the awaits involves using the dominator analysis, which we need
+        //  some time to investigate how to use. Not much, but sufficient enough for a future
+        //  improvement. See TODO inside replaceUsesWithIf.
+        auto results = executeOp.getResults();
+        // The first value is just a token.
+        std::vector<Value> bodyReturns(results.begin() + 1, results.end());
+        // TODO: Assert/FAIL that op.getResults() and bodyReturns have the same number of elements.
+        if (bodyReturns.size() == op.getResults().size()) {
+            for (auto [oldVal, newVal] : llvm::zip(op.getResults(), bodyReturns)) {
+                auto _users = oldVal.getUsers();
+                // Insert users into a vector to avoid modifying users during a loop.
+                std::vector<Operation *> users(_users.begin(), _users.end());
```

Comment:
why are both `users` and `_users` required?

Reply:
What are these users? :))

Reply:
I do not want to iterate directly on 

```c++
for (auto user : oldVal.getUsers()) {
  // ... snip ...
}
```

because we modify `user` such that it is no longer a user of `oldVal`. Depending on how the iterator is defined, this may cause a problem.

```llvm

%variable = some.op // definition of variable

%res = some.other.op %variable // operand 0 in this instruction is a use of %variable
                                                     // and this instruction is a user of %variable
```